crates/ty_python_semantic/resources/mdtest/del.md

```diff
@@ -40,22 +40,82 @@ else:
 # error: [possibly-unresolved-reference]
 reveal_type(c)  # revealed: Literal[2]
 
-d = 1
+d = [1, 2, 3]
 
 def delete():
-    # TODO: this results in `UnboundLocalError`; we should emit `unresolved-reference`
-    del d
+    del d  # error: [unresolved-reference] "Name `d` used when not defined"
 
 delete()
-reveal_type(d)  # revealed: Literal[1]
+reveal_type(d)  # revealed: list[Unknown]
+
+def delete_element():
+    # When the `del` target isn't a name, it doesn't force local resolution.
+    del d[0]
+    print(d)
 
 def delete_global():
     global d
     del d
+    # We could lint that `d` is unbound in this trivial case, but because it's global we'd need to
+    # be careful about false positives if `d` got reinitialized somehow in between the two `del`s.
+    del d
 
 delete_global()
-# The variable should have been removed, but we won't check it for now.
-reveal_type(d)  # revealed: Literal[1]
+# Again, the variable should have been removed, but we don't check it.
+reveal_type(d)  # revealed: list[Unknown]
+
+def delete_nonlocal():
+    e = 2
+
+    def delete_nonlocal_bad():
+        del e  # error: [unresolved-reference] "Name `e` used when not defined"
+
+    def delete_nonlocal_ok():
+        nonlocal e
+        del e
+        # As with `global` above, we don't track that the nonlocal `e` is unbound.
+        del e
+```
+
+## `del` forces local resolution even if it's unreachable
+
+```py
+x = 42
+
+def f():
+    print(x)  # error: [unresolved-reference] "Name `x` used when not defined"
+    if False:
+        # Assigning to `x` would have the same effect here.
+        del x
+
+    def g():
+        print(x)  # error: [unresolved-reference] "Name `x` used when not defined"
+```
+
+## `del` doesn't mark `nonlocal` variables as bound in the current scope
+
+Without a `nonlocal x` declaration in `g`, `del x` in `g` causes `print(x)` in `h` to resolve to
+`g`'s binding (i.e. an unresolved reference / unbound local error). However, with `nonlocal x` in
+`g`, `print(x)` in `h` resolves to `f`'s binding, despite the following `del`:
+
+```py
+def f():
+    x = 1
+    def g():
+        def h():
+            print(x)  # error: [unresolved-reference] "Name `x` used when not defined"
+        h()
+        if False:
+            del x  # This `del` wouldn't be allowed if it was reachable.
+
+def f():
+    x = 1
+    def g():
+        nonlocal x
+        def h():
+            print(x)  # allowed, refers to `x` in `f`
```

Comment:
If we used `reveal_type` instead of `print`, could we show (and test) that it refers to `x` in `f`, rather than just asserting that in an unchecked comment?